LoopingControl: Handle invalid start position properly (lp1942715)

Holzhaus opened the pull request:
An invalid start position resets the loop anyway.

See this for details: https://bugs.launchpad.net/mixxx/+bug/1942715

Holzhaus committed:
LoopingControl: Handle invalid start position properly (lp1942715)

An invalid start position resets the loop anyway.

See this for details: https://bugs.launchpad.net/mixxx/+bug/1942715

uklotzde commented on src/engine/controls/loopingcontrol.cpp:
```diff
@@ -965,7 +965,8 @@ void LoopingControl::slotLoopStartPos(double positionSamples) {
     m_pCOLoopStartPosition->set(position.toEngineSamplePosMaybeInvalid());
 
-    if (loopInfo.endPosition.isValid() && loopInfo.endPosition <= loopInfo.startPosition) {
+    if (loopInfo.startPosition.isValid() && loopInfo.endPosition.isValid() &&
```
We already checked for `position.isValid()` above. I don't see a reason why we cannot combine these code paths like `if (!loopInfo.startPosition.isValid() || !loopInfo.endPosition().isValid() || loopInfo.endPosition <= loopInfo.startPosition) ...`?

Holzhaus replied:
https://github.com/mixxxdj/mixxx/pull/4266/commits/d8f91a9cebe4c365f4e36d79d0e205d3e22580fb

Holzhaus committed:
LoopingControl: Combine loop reset code paths in slotLoopStartPos()

uklotzde commented on src/engine/controls/loopingcontrol.cpp:
```diff
@@ -959,18 +959,17 @@ void LoopingControl::slotLoopStartPos(double positionSamples) {
+        setLoopingEnabled(false);
     }
 
     loopInfo.seekMode = LoopSeekMode::MovedOut;
```
Let's move these lines up, closer to where `loopInfo` is initialized.

Then use `loopInfo.startPosition` instead of `position` for consistency and to reveal the dependencies. For me this is easier to follow. Less variables to remember, more concise.

Compile time enforcement: Use a local scope so that the temporary `position` becomes unavailable for the remainder of the function body.

Holzhaus replied:
https://github.com/mixxxdj/mixxx/pull/4266/commits/6fe90117fbd741517212eb080fe9829e6abe38df

Holzhaus committed:
LoopingControl: Reduce usage of `position` variable in `slotLoopStartPos()`

This sets loopInfo.startPosition first and then uses it instead of
`position` for consistency and to reveal the dependencies. Also usees a
local scope so that the temporary `position` becomes unavailable for the
remainder of the function body.

daschuer reviewed (approved):
LGTM, Thank you.
